File: JAVA/queue/ArrayQueue.java
package queue;

/**
 * 使用数组实现,队列
 */
public class ArrayQueue {
    private int[] arr;
    private int   count;

    public ArrayQueue(int size) {
        arr   = new int[size];
        count = 0;
    }

    // 将得到的val放到队列的末尾
    public void add(int val) {
        arr[count++] = val;
    }

    // 返回队列的头元素
    public int fornt() {
        return arr[0];
    }

    // 返回队列的栈顶元素,并删除栈顶元素
    public int pop() {
        int tmp = arr[0];
        count --;
        for(int i = 1; i <= count; i++) {
            arr[i -1] = arr[i];
        }

        return tmp;
    }

    // 返回栈的大小
    public int size() {
        return count;
    }

    // 判断栈是否为空
    public boolean isEmpty() {
        return size() == 0;
    }

    public static void main(String[] args) {
        int tmp = 0;
        ArrayQueue arrayQueue = new ArrayQueue(12);
        arrayQueue.add(10);
        arrayQueue.add(11);
        arrayQueue.add(12);
        arrayQueue.add(13);

        //打印队列的大小
        System.out.println("queue is size = " + arrayQueue.size());
        // 将队列头元素赋值给tmp,并删除队列头元素
        tmp = arrayQueue.pop();
        System.out.println("tmp = " + tmp);
        System.out.println("queue is size = " + arrayQueue.size());
        // 将队列头元素赋值给tmp
        tmp = arrayQueue.fornt();
        System.out.println("tmp = " + tmp);
        System.out.println("queue is size = " + arrayQueue.size());
        // 打印队列数据
        while(!arrayQueue.isEmpty()) {
            System.out.println(arrayQueue.pop());
        }
    }
}

File: C/queue/array_queue.c
#include<stdio.h>
#include<stdlib.h>

/**
 * 用数组实现队列,并且只能存储int数据
 */

// 保存数据的数组
static int *arr = NULL;
// 队列的实际大小
static int count;

// 创建队列
int create_array_queue(int size) {
    arr = (int *)malloc(size*sizeof(int));
    if (!arr) {
        printf("arr malloc error!");

        return -1;
    }
    count = 0;

    return 0;
}

// 销毁“队列”
int destory_array_queue() {
    if(arr) {
        free(arr);
        arr = NULL;
    }

    return 0;
}

//将val数据添加到的末尾
void add(int val) {
    arr[count++] = val;
}

// 返回队列开头数据
int front() {
    return arr[0];
}

// 返回并删除队列开头数据
int pop() {
    int tmp = arr[0];
    count--;
    int i = 0;
    while(i++ < count) {
        arr[i - 1] = arr[i];
    }

    return tmp;
}

// 返回队列的大小
int size() {
    return count;
}

// 返回队列是否为空
int is_empty() {
    return size() == 0;
}

void main() {
    int tmp = 0;
    
    //创建队列
    create_array_queue(12);
    //将数据存入队列中
    add(10);
    add(11);
    add(12);
    add(13);
    
    printf("queue size = %d\n", size());
    // 将队列开头数据赋值给tmp,并删除该元素
    tmp = pop();
    printf("tmp = %d\n", tmp);
    printf("queue size = %d\n", size());
    // 把队列开头的数据赋值给tmp
    tmp = front();
    printf("tmp = %d\n", tmp);
    printf("queue size = %d\n", size());
    // 答应队列
    while(!is_empty()) {
        printf("%d\n", pop());
    }

    // 摧毁队列
    destory_array_queue();
}